cookie-store/resources/cookie_helper.py

```diff
@@ -18,52 +18,56 @@
-DISALLOWED_IN_HEADER_RE = re.compile(r'[\0-\x1f\x7f]');
+DISALLOWED_IN_COOKIE_NAME_RE = re.compile(r'[;\0-\x1f\x7f]')
+DISALLOWED_IN_HEADER_RE = re.compile(r'[\0-\x1f\x7f]')
 
 # Ensure common charset names do not end up with different
 # capitalization or punctuation
 CHARSET_OVERRIDES = {
     encodings.codecs.lookup(charset).name: charset
-    for charset in ('utf-8', 'iso-8859-1',)
+    for charset in (u'utf-8', u'iso-8859-1',)
 }
 
 def main(request, response):
   assert request.method in (
-      'GET',
-      'POST',
-  ), 'request method was neither GET nor POST: %r' % request.method
-  qd = (request.url.split('#')[0].split('?', 1) + [''])[1]
-  if request.method == 'POST':
-    qd += '&' + request.body
-  args = cgi.parse_qs(qd, keep_blank_values = True)
-  charset = encodings.codecs.lookup(args.get('charset', ['utf-8'])[-1]).name
+      u'GET',
+      u'POST',
+  ), u'request method was neither GET nor POST: %r' % request.method
+  qd = (request.url.encode("utf-8").split(b'#')[0].split(b'?', 1) + [b''])[1]
+  if request.method == u'POST':
+    qd += b'&' + request.body
+  args = parse_qs(qd, keep_blank_values = True)
+  charset = encodings.codecs.lookup(args.get(u'charset', [u'utf-8'])[-1]).name
   charset = CHARSET_OVERRIDES.get(charset, charset)
-  headers = [('content-type', 'text/plain; charset=' + charset)]
+  headers = [(b'content-type', b'text/plain; charset=' + isomorphic_encode(charset))]
   body = []
-  if request.method == 'POST':
-    for set_cookie in args.get('set-cookie', []):
-      if '=' in set_cookie.split(';', 1)[0]:
-        name, rest = set_cookie.split('=', 1)
+  if request.method == u'POST':
+    for set_cookie in args.get(b'set-cookie', []):
+      if b'=' in set_cookie.split(b';', 1)[0]:
+        name, rest = set_cookie.split(b'=', 1)
         assert re.search(
             DISALLOWED_IN_COOKIE_NAME_RE,
-            name
-        ) is None, 'name had disallowed characters: %r' % name
+            isomorphic_decode(name)
+        ) is None, b'name had disallowed characters: %r' % name
       else:
         rest = set_cookie
       assert re.search(
           DISALLOWED_IN_HEADER_RE,
-          rest
-      ) is None, 'rest had disallowed characters: %r' % rest
-      headers.append(('set-cookie', set_cookie))
-      body.append('set-cookie=' + urllib.quote(set_cookie, ''))
+          isomorphic_decode(rest)
+      ) is None, b'rest had disallowed characters: %r' % rest
+      headers.append((b'set-cookie', set_cookie))
+      body.append(u'set-cookie=' + quote(set_cookie, b''))
   else:
-    cookie = request.headers.get('cookie')
+    cookie = request.headers.get(b'cookie')
     if cookie is not None:
-      body.append('cookie=' + urllib.quote(cookie, ''))
-  body = '\r\n'.join(body)
-  headers.append(('content-length', str(len(body))))
+      body.append(u'cookie=' + quote(cookie, b''))
+  body = u'\r\n'.join(body)
+  headers.append((b'content-length', isomorphic_encode(str(len(body)))))
```

Comment:
I wonder if you could (not sure if it'd work)

```suggestion
  headers.append((b'content-length', len(body)))
```